DynamicActiveQuery.php

```diff
@@ -28,6 +30,43 @@ class DynamicActiveQuery extends ActiveQuery
     private $dynamicColumn;
 
     /**
+     * Convert index value to closure, that will get decoded dynamic attribute, in case if indexing attribute is dynamic
+     * @param callable|string $column
+     * @return $this
+     */
+    public function indexBy($column)
+    {
+        if (!$this->asArray) {
+            return parent::indexBy($column);
+        }
+
+        /** @var \yii\db\ActiveRecord $modelClass */
+        $modelClass = $this->modelClass;
+        $this->indexBy = function ($row) use ($column, $modelClass) {
+            if (isset($row[$column])) {
+                return $row[$column];
+            }
+
+            if (!method_exists($modelClass, 'dynamicColumn')) {
+                throw new UnknownMethodException("Method {$modelClass}::dynamicColumn() does not exist");
+            }
```

Comment:
i wasn't able to trigger this exception. DynamicActiveRecord threw an exception first. that does not prove that we do not need this exception but i would like to have a test that shows it being used.